packages/core/feature-flags/core-feature-flags-server-internal/src/feature_flags_service.ts

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { CoreContext } from '@kbn/core-base-server-internal';
+import type {
+  EvaluationContext,
+  FeatureFlagsSetup,
+  FeatureFlagsStart,
+  MultiContextEvaluationContext,
+} from '@kbn/core-feature-flags-server';
+import type { Logger } from '@kbn/logging';
+import apm from 'elastic-apm-node';
+import {
+  type Client,
+  OpenFeature,
+  ServerProviderEvents,
+  NOOP_PROVIDER,
+} from '@openfeature/server-sdk';
+import deepMerge from 'deepmerge';
+import { filter, switchMap, startWith, Subject } from 'rxjs';
+import { type FeatureFlagsConfig, featureFlagsConfig } from './feature_flags_config';
+
+/**
+ * Core-internal contract for the setup lifecycle step.
+ * @private
+ */
+export interface InternalFeatureFlagsSetup extends FeatureFlagsSetup {
+  /**
+   * Used by the rendering service to share the overrides with the service on the browser side.
+   */
+  getOverrides: () => Record<string, unknown>;
+}
+
+/**
+ * The server-side Feature Flags Service
+ * @private
+ */
+export class FeatureFlagsService {
+  private readonly featureFlagsClient: Client;
+  private readonly logger: Logger;
+  private overrides: Record<string, unknown> = {};
+  private context: MultiContextEvaluationContext = { kind: 'multi' };
+
+  /**
+   * The core service's constructor
+   * @param core {@link CoreContext}
+   */
+  constructor(private readonly core: CoreContext) {
+    this.logger = core.logger.get('feature-flags-service');
+    this.featureFlagsClient = OpenFeature.getClient();
+    OpenFeature.setLogger(this.logger.get('open-feature'));
```

Comment:
I tried running all services in debug level, and log entries look safe to me:

<details><summary>With an invalid SDK key (server side)</summary>
<pre>
[DEBUG][plugins.featureFlagsExample] "/Users/afharo/Developer/elastic/kibana/examples/feature_flags_example/server" does not export "config". {"service":{"node":{"roles":["background_tasks","ui"]}}}
[DEBUG][plugins.cloudExperiments] Initializing plugin {"service":{"node":{"roles":["background_tasks","ui"]}}}
[DEBUG][plugins.cloudExperiments.launch-darkly] Started EventProcessor. {"service":{"node":{"roles":["background_tasks","ui"]}}}
[DEBUG][plugins.featureFlagsExample] Initializing plugin {"service":{"node":{"roles":["background_tasks","ui"]}}}
[INFO ][plugins.featureFlagsExample] The observed feature flags are:
      - example-boolean: false
      - example-string: red
      - example-number: 1
       {"service":{"node":{"roles":["background_tasks","ui"]}}}
[INFO ][plugins.featureFlagsExample] The feature flags are:
      - example-boolean: false
      - example-string: white
      - example-number: 1
       {"service":{"node":{"roles":["background_tasks","ui"]}}}
[ERROR][plugins.cloudExperiments.launch-darkly] Authentication failed. Double check your SDK key. {"service":{"node":{"roles":["background_tasks","ui"]}}}
[ERROR][plugins.cloudExperiments.launch-darkly] Received error 401 (invalid SDK key) for streaming request - giving up permanently {"service":{"node":{"roles":["background_tasks","ui"]}}}
[ERROR][feature-flags-service.open-feature] Error during provider initialization: {"service":{"node":{"roles":["background_tasks","ui"]}}}
</pre>
</details> 

<details><summary>With an invalid Client ID (browser side)</summary>
<pre>
feature_flags_service.ts:56 warn: [LaunchDarkly] The waitForInitialization function was called without a timeout specified. In a future version a default timeout will be applied.
ldclient.cjs.js:1 debug: [LaunchDarkly] sending diagnostic event (diagnostic-combined)
ldclient.cjs.js:1 debug: [LaunchDarkly] enqueueing "identify" event
ldclient.cjs.js:1 debug: [LaunchDarkly] polling for feature flags at https://app.launchdarkly.com/sdk/evalx/668ea35e5dd98d109e1bc7d/contexts/eyJhbm9ueW1vdXMiOnRydWUsImtleSI6ImJmYjBiMjYwLTQzOGUtMTFlZi05OGJmLTFmNTk2ODdjY2ZmOCJ9
feature_flags_service.ts:56 GET https://app.launchdarkly.com/sdk/goals/668ea35e5dd98d109e1bc7d 404 (Not Found)
(anonymous) @ xhr-patch.js:94
proto.<computed> @ patch-utils.js:54
(anonymous) @ ldclient.cjs.js:1
Ot @ ldclient.cjs.js:1
n.httpRequest @ ldclient.cjs.js:1
l @ ldclient.cjs.js:1
Qe.c.fetchJSON @ ldclient.cjs.js:1
At @ ldclient.cjs.js:1
Nt @ ldclient.cjs.js:1
(anonymous) @ index.cjs.js:278
(anonymous) @ index.cjs.js:46
__awaiter @ index.cjs.js:42
initialize @ index.cjs.js:275
setAwaitableProvider @ index.js:595
(anonymous) @ index.js:681
(anonymous) @ index.js:69
__async @ index.js:53
setProviderAndWait @ index.js:670
setProvider @ feature_flags_service.ts:56
setup @ plugin.ts:50
setup @ plugin.ts:48
setup @ plugins_service.ts:71
setup @ core_system.ts:262
await in setup
__kbnBootstrap__ @ kbn_bootstrap.ts:34
await in __kbnBootstrap__
(anonymous) @ bootstrap.js:136
innerCb @ bootstrap.js:115
load
loadScript @ bootstrap.js:105
(anonymous) @ bootstrap.js:124
load @ bootstrap.js:111
window.onload @ bootstrap.js:133
load
(anonymous) @ bootstrap.js:47
Show 18 more frames
Show less
ldclient.cjs.js:1 error: [LaunchDarkly] Environment not found. Double check that you specified a valid environment/client-side ID. Please see https://docs.launchdarkly.com/sdk/client-side/javascript#initialize-the-client for instructions on SDK initialization.
ldclient.cjs.js:1 debug: [LaunchDarkly] polling for feature flags at https://app.launchdarkly.com/sdk/evalx/668ea35e5dd98d109e1bc7d/contexts/eyJraW5kIjoibXVsdGkiLCJraWJhbmEiOnsia2V5IjoiMTIzNCIsIm9mZmVyaW5nIjoidHJhZGl0aW9uYWwiLCJ2ZXJzaW9uIjoiOC4xNi4wIiwiYnVpbGRfbnVtIjo5MDA3MTk5MjU0NzQwOTkxLCJidWlsZF9zaGEiOiJYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYIiwiYnVpbGRfc2hhX3Nob3J0IjoiWFhYWFhYWFhYWFhYIn0sIm9yZ2FuaXphdGlvbiI6eyJrZXkiOiIxMjM0NSIsImlzX2VsYXN0aWNfc3RhZmYiOnRydWV9fQ
feature_flags_service.ts:56 POST https://events.launchdarkly.com/events/diagnostic/668ea35e5dd98d109e1bc7d 404 (Not Found)
(anonymous) @ xhr-patch.js:94
proto.<computed> @ patch-utils.js:54
(anonymous) @ ldclient.cjs.js:1
Ot @ ldclient.cjs.js:1
n.httpRequest @ ldclient.cjs.js:1
t @ ldclient.cjs.js:1
o.sendEvents @ ldclient.cjs.js:1
h @ ldclient.cjs.js:1
y @ ldclient.cjs.js:1
(anonymous) @ ldclient.cjs.js:1
(anonymous) @ ldclient.cjs.js:1
Promise.then
(anonymous) @ ldclient.cjs.js:1
DiagnosticsManager.p.start @ ldclient.cjs.js:1
start @ ldclient.cjs.js:1
Nt @ ldclient.cjs.js:1
(anonymous) @ index.cjs.js:278
(anonymous) @ index.cjs.js:46
__awaiter @ index.cjs.js:42
initialize @ index.cjs.js:275
setAwaitableProvider @ index.js:595
(anonymous) @ index.js:681
(anonymous) @ index.js:69
__async @ index.js:53
setProviderAndWait @ index.js:670
setProvider @ feature_flags_service.ts:56
setup @ plugin.ts:50
setup @ plugin.ts:48
setup @ plugins_service.ts:71
setup @ core_system.ts:262
await in setup
__kbnBootstrap__ @ kbn_bootstrap.ts:34
await in __kbnBootstrap__
(anonymous) @ bootstrap.js:136
innerCb @ bootstrap.js:115
load
loadScript @ bootstrap.js:105
(anonymous) @ bootstrap.js:124
load @ bootstrap.js:111
window.onload @ bootstrap.js:133
load
(anonymous) @ bootstrap.js:47
Show 24 more frames
Show less
feature_flags_service.ts:155 GET https://app.launchdarkly.com/sdk/evalx/668ea35e5dd98d109e1bc7d/contexts/eyJraW5kIjoibXVsdGkiLCJraWJhbmEiOnsia2V5IjoiMTIzNCIsIm9mZmVyaW5nIjoidHJhZGl0aW9uYWwiLCJ2ZXJzaW9uIjoiOC4xNi4wIiwiYnVpbGRfbnVtIjo5MDA3MTk5MjU0NzQwOTkxLCJidWlsZF9zaGEiOiJYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYIiwiYnVpbGRfc2hhX3Nob3J0IjoiWFhYWFhYWFhYWFhYIn0sIm9yZ2FuaXphdGlvbiI6eyJrZXkiOiIxMjM0NSIsImlzX2VsYXN0aWNfc3RhZmYiOnRydWV9fQ 404 (Not Found)
(anonymous) @ xhr-patch.js:94
proto.<computed> @ patch-utils.js:54
(anonymous) @ ldclient.cjs.js:1
Ot @ ldclient.cjs.js:1
n.httpRequest @ ldclient.cjs.js:1
l @ ldclient.cjs.js:1
Qe.c.fetchFlagSettings @ ldclient.cjs.js:1
(anonymous) @ ldclient.cjs.js:1
Promise.then
identify @ ldclient.cjs.js:1
(anonymous) @ index.cjs.js:311
(anonymous) @ index.cjs.js:46
__awaiter @ index.cjs.js:42
onContextChange @ index.cjs.js:309
(anonymous) @ index.js:811
(anonymous) @ index.js:69
__async @ index.js:53
runProviderContextChangeHandler @ index.js:806
(anonymous) @ index.js:728
(anonymous) @ index.js:728
(anonymous) @ index.js:69
__async @ index.js:53
setContext @ index.js:702
appendContext @ feature_flags_service.ts:155
appendContext @ feature_flags_service.ts:58
(anonymous) @ initialize_metadata.ts:40
doInnerSub @ mergeInternals.js:19
outerNext @ mergeInternals.js:14
OperatorSubscriber._this._next @ OperatorSubscriber.js:15
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber.next @ Subscriber.js:34
(anonymous) @ Subject.js:41
errorContext @ errorContext.js:19
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subject.js.Subject.next @ Subject.js:31
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/ReplaySubject.js.ReplaySubject.next @ ReplaySubject.js:28
next @ share.js:46
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.ConsumerObserver.next @ Subscriber.js:96
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber._next @ Subscriber.js:63
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber.next @ Subscriber.js:34
(anonymous) @ map.js:7
OperatorSubscriber._this._next @ OperatorSubscriber.js:15
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber.next @ Subscriber.js:34
(anonymous) @ distinct.js:12
OperatorSubscriber._this._next @ OperatorSubscriber.js:15
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber.next @ Subscriber.js:34
emit @ debounceTime.js:16
emitWhenIdle @ debounceTime.js:27
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/scheduler/AsyncAction.js.AsyncAction._execute @ AsyncAction.js:62
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/scheduler/AsyncAction.js.AsyncAction.execute @ AsyncAction.js:50
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/scheduler/AsyncScheduler.js.AsyncScheduler.flush @ AsyncScheduler.js:21
setInterval
setInterval @ intervalProvider.js:12
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/scheduler/AsyncAction.js.AsyncAction.requestAsyncId @ AsyncAction.js:33
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/scheduler/AsyncAction.js.AsyncAction.schedule @ AsyncAction.js:28
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Scheduler.js.Scheduler.schedule @ Scheduler.js:10
(anonymous) @ debounceTime.js:33
OperatorSubscriber._this._next @ OperatorSubscriber.js:15
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber.next @ Subscriber.js:34
(anonymous) @ filter.js:6
OperatorSubscriber._this._next @ OperatorSubscriber.js:15
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber.next @ Subscriber.js:34
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/BehaviorSubject.js.BehaviorSubject._subscribe @ BehaviorSubject.js:19
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Observable.js.Observable._trySubscribe @ Observable.js:38
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subject.js.Subject._trySubscribe @ Subject.js:95
(anonymous) @ Observable.js:32
errorContext @ errorContext.js:19
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Observable.js.Observable.subscribe @ Observable.js:23
(anonymous) @ filter.js:6
(anonymous) @ lift.js:10
(anonymous) @ Observable.js:27
errorContext @ errorContext.js:19
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Observable.js.Observable.subscribe @ Observable.js:23
(anonymous) @ debounceTime.js:29
(anonymous) @ lift.js:10
(anonymous) @ Observable.js:27
errorContext @ errorContext.js:19
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Observable.js.Observable.subscribe @ Observable.js:23
(anonymous) @ distinct.js:8
(anonymous) @ lift.js:10
(anonymous) @ Observable.js:27
errorContext @ errorContext.js:19
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Observable.js.Observable.subscribe @ Observable.js:23
(anonymous) @ map.js:6
(anonymous) @ lift.js:10
(anonymous) @ Observable.js:27
errorContext @ errorContext.js:19
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Observable.js.Observable.subscribe @ Observable.js:23
(anonymous) @ share.js:60
(anonymous) @ lift.js:10
(anonymous) @ Observable.js:27
errorContext @ errorContext.js:19
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Observable.js.Observable.subscribe @ Observable.js:23
mergeInternals @ mergeInternals.js:53
(anonymous) @ mergeMap.js:14
(anonymous) @ lift.js:10
(anonymous) @ Observable.js:27
errorContext @ errorContext.js:19
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Observable.js.Observable.subscribe @ Observable.js:23
initializeMetadata @ initialize_metadata.ts:44
setup @ plugin.ts:41
setup @ plugin.ts:48
setup @ plugins_service.ts:71
setup @ core_system.ts:262
await in setup
__kbnBootstrap__ @ kbn_bootstrap.ts:34
await in __kbnBootstrap__
(anonymous) @ bootstrap.js:136
innerCb @ bootstrap.js:115
load
loadScript @ bootstrap.js:105
(anonymous) @ bootstrap.js:124
load @ bootstrap.js:111
window.onload @ bootstrap.js:133
load
(anonymous) @ bootstrap.js:47
Show 92 more frames
Show less
ldclient.cjs.js:1 error: [LaunchDarkly] Environment not found. Double check that you specified a valid environment/client-side ID. Please see https://docs.launchdarkly.com/sdk/client-side/javascript#initialize-the-client for instructions on SDK initialization.
console_appender.ts:31 [2024-08-19T19:42:43.614+02:00][ERROR][feature-flags-service.open-feature] Error running launchdarkly-client-provider's context change handler: Environment not found. Double check that you specified a valid environment/client-side ID. Please see https://docs.launchdarkly.com/sdk/client-side/javascript#initialize-the-client for instructions on SDK initialization.
feature_flags_service.ts:56 GET https://app.launchdarkly.com/sdk/evalx/668ea35e5dd98d109e1bc7d/contexts/eyJhbm9ueW1vdXMiOnRydWUsImtleSI6ImJmYjBiMjYwLTQzOGUtMTFlZi05OGJmLTFmNTk2ODdjY2ZmOCJ9 404 (Not Found)
(anonymous) @ xhr-patch.js:94
proto.<computed> @ patch-utils.js:54
(anonymous) @ ldclient.cjs.js:1
Ot @ ldclient.cjs.js:1
n.httpRequest @ ldclient.cjs.js:1
l @ ldclient.cjs.js:1
Qe.c.fetchFlagSettings @ ldclient.cjs.js:1
(anonymous) @ ldclient.cjs.js:1
Promise.then
(anonymous) @ ldclient.cjs.js:1
initialize @ ldclient.cjs.js:1
Nt @ ldclient.cjs.js:1
(anonymous) @ index.cjs.js:278
(anonymous) @ index.cjs.js:46
__awaiter @ index.cjs.js:42
initialize @ index.cjs.js:275
setAwaitableProvider @ index.js:595
(anonymous) @ index.js:681
(anonymous) @ index.js:69
__async @ index.js:53
setProviderAndWait @ index.js:670
setProvider @ feature_flags_service.ts:56
setup @ plugin.ts:50
setup @ plugin.ts:48
setup @ plugins_service.ts:71
setup @ core_system.ts:262
await in setup
__kbnBootstrap__ @ kbn_bootstrap.ts:34
await in __kbnBootstrap__
(anonymous) @ bootstrap.js:136
innerCb @ bootstrap.js:115
load
loadScript @ bootstrap.js:105
(anonymous) @ bootstrap.js:124
load @ bootstrap.js:111
window.onload @ bootstrap.js:133
load
(anonymous) @ bootstrap.js:47
Show 20 more frames
Show less
ldclient.cjs.js:1 error: [LaunchDarkly] Environment not found. Double check that you specified a valid environment/client-side ID. Please see https://docs.launchdarkly.com/sdk/client-side/javascript#initialize-the-client for instructions on SDK initialization.
ldclient.cjs.js:1 debug: [LaunchDarkly] polling for feature flags at https://app.launchdarkly.com/sdk/evalx/668ea35e5dd98d109e1bc7d/contexts/eyJraW5kIjoibXVsdGkiLCJraWJhbmEiOnsia2V5IjoiMTIzNCIsIm9mZmVyaW5nIjoidHJhZGl0aW9uYWwiLCJ2ZXJzaW9uIjoiOC4xNi4wIiwiYnVpbGRfbnVtIjo5MDA3MTk5MjU0NzQwOTkxLCJidWlsZF9zaGEiOiJYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYIiwiYnVpbGRfc2hhX3Nob3J0IjoiWFhYWFhYWFhYWFhYIiwiaGFzX2RhdGEiOmZhbHNlfSwib3JnYW5pemF0aW9uIjp7ImtleSI6IjEyMzQ1IiwiaXNfZWxhc3RpY19zdGFmZiI6dHJ1ZX19
feature_flags_service.ts:155 GET https://app.launchdarkly.com/sdk/evalx/668ea35e5dd98d109e1bc7d/contexts/eyJraW5kIjoibXVsdGkiLCJraWJhbmEiOnsia2V5IjoiMTIzNCIsIm9mZmVyaW5nIjoidHJhZGl0aW9uYWwiLCJ2ZXJzaW9uIjoiOC4xNi4wIiwiYnVpbGRfbnVtIjo5MDA3MTk5MjU0NzQwOTkxLCJidWlsZF9zaGEiOiJYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYIiwiYnVpbGRfc2hhX3Nob3J0IjoiWFhYWFhYWFhYWFhYIiwiaGFzX2RhdGEiOmZhbHNlfSwib3JnYW5pemF0aW9uIjp7ImtleSI6IjEyMzQ1IiwiaXNfZWxhc3RpY19zdGFmZiI6dHJ1ZX19 404 (Not Found)
(anonymous) @ xhr-patch.js:94
proto.<computed> @ patch-utils.js:54
(anonymous) @ ldclient.cjs.js:1
Ot @ ldclient.cjs.js:1
n.httpRequest @ ldclient.cjs.js:1
l @ ldclient.cjs.js:1
Qe.c.fetchFlagSettings @ ldclient.cjs.js:1
(anonymous) @ ldclient.cjs.js:1
Promise.then
identify @ ldclient.cjs.js:1
(anonymous) @ index.cjs.js:311
(anonymous) @ index.cjs.js:46
__awaiter @ index.cjs.js:42
onContextChange @ index.cjs.js:309
(anonymous) @ index.js:811
(anonymous) @ index.js:69
__async @ index.js:53
runProviderContextChangeHandler @ index.js:806
(anonymous) @ index.js:728
(anonymous) @ index.js:728
(anonymous) @ index.js:69
__async @ index.js:53
setContext @ index.js:702
appendContext @ feature_flags_service.ts:155
appendContext @ feature_flags_service.ts:58
(anonymous) @ initialize_metadata.ts:40
doInnerSub @ mergeInternals.js:19
outerNext @ mergeInternals.js:14
OperatorSubscriber._this._next @ OperatorSubscriber.js:15
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber.next @ Subscriber.js:34
(anonymous) @ Subject.js:41
errorContext @ errorContext.js:19
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subject.js.Subject.next @ Subject.js:31
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/ReplaySubject.js.ReplaySubject.next @ ReplaySubject.js:28
next @ share.js:46
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.ConsumerObserver.next @ Subscriber.js:96
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber._next @ Subscriber.js:63
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber.next @ Subscriber.js:34
(anonymous) @ map.js:7
OperatorSubscriber._this._next @ OperatorSubscriber.js:15
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber.next @ Subscriber.js:34
(anonymous) @ distinct.js:12
OperatorSubscriber._this._next @ OperatorSubscriber.js:15
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber.next @ Subscriber.js:34
emit @ debounceTime.js:16
emitWhenIdle @ debounceTime.js:27
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/scheduler/AsyncAction.js.AsyncAction._execute @ AsyncAction.js:62
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/scheduler/AsyncAction.js.AsyncAction.execute @ AsyncAction.js:50
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/scheduler/AsyncScheduler.js.AsyncScheduler.flush @ AsyncScheduler.js:21
setInterval
setInterval @ intervalProvider.js:12
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/scheduler/AsyncAction.js.AsyncAction.requestAsyncId @ AsyncAction.js:33
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/scheduler/AsyncAction.js.AsyncAction.schedule @ AsyncAction.js:28
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Scheduler.js.Scheduler.schedule @ Scheduler.js:10
(anonymous) @ debounceTime.js:33
OperatorSubscriber._this._next @ OperatorSubscriber.js:15
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber.next @ Subscriber.js:34
(anonymous) @ filter.js:6
OperatorSubscriber._this._next @ OperatorSubscriber.js:15
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber.next @ Subscriber.js:34
(anonymous) @ Subject.js:41
errorContext @ errorContext.js:19
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subject.js.Subject.next @ Subject.js:31
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/BehaviorSubject.js.BehaviorSubject.next @ BehaviorSubject.js:31
(anonymous) @ metadata_service.ts:90
await in (anonymous)
(anonymous) @ exhaustMap.js:21
OperatorSubscriber._this._next @ OperatorSubscriber.js:15
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/Subscriber.js.Subscriber.next @ Subscriber.js:34
(anonymous) @ timer.js:25
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/scheduler/AsyncAction.js.AsyncAction._execute @ AsyncAction.js:62
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/scheduler/AsyncAction.js.AsyncAction.execute @ AsyncAction.js:50
__kbnSharedDeps_npm__.../../node_modules/rxjs/dist/esm5/internal/scheduler/AsyncScheduler.js.AsyncScheduler.flush @ AsyncScheduler.js:21
Show 66 more frames
Show less
ldclient.cjs.js:1 error: [LaunchDarkly] Environment not found. Double check that you specified a valid environment/client-side ID. Please see https://docs.launchdarkly.com/sdk/client-side/javascript#initialize-the-client for instructions on SDK initialization.
console_appender.ts:31 [2024-08-19T19:42:44.417+02:00][ERROR][feature-flags-service.open-feature] Error running launchdarkly-client-provider's context change handler: Environment not found. Double check that you specified a valid environment/client-side ID. Please see https://docs.launchdarkly.com/sdk/client-side/javascript#initialize-the-client for instructions on SDK initialization.
ldclient.cjs.js:1 debug: [LaunchDarkly] sending 1 events
xhr-patch.js:94 POST https://events.launchdarkly.com/events/bulk/668ea35e5dd98d109e1bc7d 404 (Not Found)
(anonymous) @ xhr-patch.js:94
proto.<computed> @ patch-utils.js:54
(anonymous) @ ldclient.cjs.js:1
Ot @ ldclient.cjs.js:1
n.httpRequest @ ldclient.cjs.js:1
t @ ldclient.cjs.js:1
o.sendEvents @ ldclient.cjs.js:1
Ue.a.flush @ ldclient.cjs.js:1
e @ ldclient.cjs.js:1
Show 9 more frames
Show less
ldclient.cjs.js:1 error: [LaunchDarkly] Received error 404 for event posting - giving up permanently
</pre>
</details> 

<details><summary>With a valid SDK key (server side)</summary>
<pre>
[DEBUG][plugins.featureFlagsExample] "/Users/afharo/Developer/elastic/kibana/examples/feature_flags_example/server" does not export "config". {"service":{"node":{"roles":["background_tasks","ui"]}}}
[DEBUG][plugins.cloudExperiments] Initializing plugin {"service":{"node":{"roles":["background_tasks","ui"]}}}
[DEBUG][plugins.cloudExperiments.launch-darkly] Started EventProcessor. {"service":{"node":{"roles":["background_tasks","ui"]}}}
[DEBUG][plugins.featureFlagsExample] Initializing plugin {"service":{"node":{"roles":["background_tasks","ui"]}}}
[INFO ][plugins.featureFlagsExample] The observed feature flags are:
      - example-boolean: false
      - example-string: red
      - example-number: 1
       {"service":{"node":{"roles":["background_tasks","ui"]}}}
[INFO ][plugins.featureFlagsExample] The feature flags are:
      - example-boolean: false
      - example-string: white
      - example-number: 1
       {"service":{"node":{"roles":["background_tasks","ui"]}}}
[INFO ][plugins.cloudExperiments.launch-darkly] Opened LaunchDarkly stream connection {"service":{"node":{"roles":["background_tasks","ui"]}}}
[DEBUG][plugins.cloudExperiments.launch-darkly] Received put event {"service":{"node":{"roles":["background_tasks","ui"]}}}
[DEBUG][plugins.cloudExperiments.launch-darkly] Initializing all data {"service":{"node":{"roles":["background_tasks","ui"]}}}
</pre>
</details> 

<details><summary>With a valid client ID (browser side)</summary>
<pre>
feature_flags_service.ts:56 warn: [LaunchDarkly] The waitForInitialization function was called without a timeout specified. In a future version a default timeout will be applied.
ldclient.cjs.js:1 debug: [LaunchDarkly] sending diagnostic event (diagnostic-combined)
ldclient.cjs.js:1 debug: [LaunchDarkly] enqueueing "identify" event
ldclient.cjs.js:1 debug: [LaunchDarkly] polling for feature flags at https://app.launchdarkly.com/sdk/evalx/668ea35e5dd98d109e1bc7d0/contexts/eyJhbm9ueW1vdXMiOnRydWUsImtleSI6ImJmYjBiMjYwLTQzOGUtMTFlZi05OGJmLTFmNTk2ODdjY2ZmOCJ9
ldclient.cjs.js:1 debug: [LaunchDarkly] polling for feature flags at https://app.launchdarkly.com/sdk/evalx/668ea35e5dd98d109e1bc7d0/contexts/eyJraW5kIjoibXVsdGkiLCJraWJhbmEiOnsia2V5IjoiMTIzNCIsIm9mZmVyaW5nIjoidHJhZGl0aW9uYWwiLCJ2ZXJzaW9uIjoiOC4xNi4wIiwiYnVpbGRfbnVtIjo5MDA3MTk5MjU0NzQwOTkxLCJidWlsZF9zaGEiOiJYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYIiwiYnVpbGRfc2hhX3Nob3J0IjoiWFhYWFhYWFhYWFhYIn0sIm9yZ2FuaXphdGlvbiI6eyJrZXkiOiIxMjM0NSIsImlzX2VsYXN0aWNfc3RhZmYiOnRydWV9fQ
ldclient.cjs.js:1 info: [LaunchDarkly] LaunchDarkly client initialized
ldclient.cjs.js:1 info: [LaunchDarkly] Opening stream connection to https://clientstream.launchdarkly.com/eval/668ea35e5dd98d109e1bc7d0/eyJhbm9ueW1vdXMiOnRydWUsImtleSI6ImJmYjBiMjYwLTQzOGUtMTFlZi05OGJmLTFmNTk2ODdjY2ZmOCJ9
ldclient.cjs.js:1 debug: [LaunchDarkly] enqueueing "identify" event
ldclient.cjs.js:1 info: [LaunchDarkly] Closing stream connection
ldclient.cjs.js:1 info: [LaunchDarkly] Opening stream connection to https://clientstream.launchdarkly.com/eval/668ea35e5dd98d109e1bc7d0/eyJraW5kIjoibXVsdGkiLCJraWJhbmEiOnsia2V5IjoiMTIzNCIsIm9mZmVyaW5nIjoidHJhZGl0aW9uYWwiLCJ2ZXJzaW9uIjoiOC4xNi4wIiwiYnVpbGRfbnVtIjo5MDA3MTk5MjU0NzQwOTkxLCJidWlsZF9zaGEiOiJYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYIiwiYnVpbGRfc2hhX3Nob3J0IjoiWFhYWFhYWFhYWFhYIn0sIm9yZ2FuaXphdGlvbiI6eyJrZXkiOiIxMjM0NSIsImlzX2VsYXN0aWNfc3RhZmYiOnRydWV9fQ
ldclient.cjs.js:1 debug: [LaunchDarkly] received streaming update for all flags
ldclient.cjs.js:1 debug: [LaunchDarkly] polling for feature flags at https://app.launchdarkly.com/sdk/evalx/668ea35e5dd98d109e1bc7d0/contexts/eyJraW5kIjoibXVsdGkiLCJraWJhbmEiOnsia2V5IjoiMTIzNCIsIm9mZmVyaW5nIjoidHJhZGl0aW9uYWwiLCJ2ZXJzaW9uIjoiOC4xNi4wIiwiYnVpbGRfbnVtIjo5MDA3MTk5MjU0NzQwOTkxLCJidWlsZF9zaGEiOiJYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYIiwiYnVpbGRfc2hhX3Nob3J0IjoiWFhYWFhYWFhYWFhYIiwiaGFzX2RhdGEiOmZhbHNlfSwib3JnYW5pemF0aW9uIjp7ImtleSI6IjEyMzQ1IiwiaXNfZWxhc3RpY19zdGFmZiI6dHJ1ZX19
ldclient.cjs.js:1 debug: [LaunchDarkly] enqueueing "identify" event
ldclient.cjs.js:1 debug: [LaunchDarkly] enqueueing "feature" event
ldclient.cjs.js:1 info: [LaunchDarkly] Closing stream connection
ldclient.cjs.js:1 info: [LaunchDarkly] Opening stream connection to https://clientstream.launchdarkly.com/eval/668ea35e5dd98d109e1bc7d0/eyJraW5kIjoibXVsdGkiLCJraWJhbmEiOnsia2V5IjoiMTIzNCIsIm9mZmVyaW5nIjoidHJhZGl0aW9uYWwiLCJ2ZXJzaW9uIjoiOC4xNi4wIiwiYnVpbGRfbnVtIjo5MDA3MTk5MjU0NzQwOTkxLCJidWlsZF9zaGEiOiJYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYIiwiYnVpbGRfc2hhX3Nob3J0IjoiWFhYWFhYWFhYWFhYIiwiaGFzX2RhdGEiOmZhbHNlfSwib3JnYW5pemF0aW9uIjp7ImtleSI6IjEyMzQ1IiwiaXNfZWxhc3RpY19zdGFmZiI6dHJ1ZX19
ldclient.cjs.js:1 debug: [LaunchDarkly] received streaming update for all flags
ldclient.cjs.js:1 debug: [LaunchDarkly] sending 4 events
</pre>
</details> 

<details><summary>Intentionally providing a malformed context (missing mandatory `key`) (browser only: server didn't complain)</summary>
<pre>
feature_flags_service.ts:56 warn: [LaunchDarkly] The waitForInitialization function was called without a timeout specified. In a future version a default timeout will be applied.
ldclient.cjs.js:1 debug: [LaunchDarkly] enqueueing "identify" event
ldclient.cjs.js:1 debug: [LaunchDarkly] polling for feature flags at https://app.launchdarkly.com/sdk/evalx/668ea35e5dd98d109e1bc7d0/contexts/eyJhbm9ueW1vdXMiOnRydWUsImtleSI6ImJmYjBiMjYwLTQzOGUtMTFlZi05OGJmLTFmNTk2ODdjY2ZmOCJ9
2feature_flags_service.ts:155 error: [LaunchDarkly] The EvaluationContext must contain either a 'targetingKey' or a 'key' and the type must be a string.
ldclient.cjs.js:1 error: [LaunchDarkly] Invalid context specified. Please see https://docs.launchdarkly.com/sdk/client-side/javascript#initialize-the-client for instructions on SDK initialization.
console_appender.ts:31 [2024-08-19T19:48:16.369+02:00][ERROR][feature-flags-service.open-feature] Error running launchdarkly-client-provider's context change handler: Invalid context specified. Please see https://docs.launchdarkly.com/sdk/client-side/javascript#initialize-the-client for instructions on SDK initialization.
ldclient.cjs.js:1 info: [LaunchDarkly] LaunchDarkly client initialized
ldclient.cjs.js:1 info: [LaunchDarkly] Opening stream connection to https://clientstream.launchdarkly.com/eval/668ea35e5dd98d109e1bc7d0/eyJhbm9ueW1vdXMiOnRydWUsImtleSI6ImJmYjBiMjYwLTQzOGUtMTFlZi05OGJmLTFmNTk2ODdjY2ZmOCJ9
ldclient.cjs.js:1 debug: [LaunchDarkly] received streaming update for all flags
2feature_flags_service.ts:155 error: [LaunchDarkly] The EvaluationContext must contain either a 'targetingKey' or a 'key' and the type must be a string.
ldclient.cjs.js:1 error: [LaunchDarkly] Invalid context specified. Please see https://docs.launchdarkly.com/sdk/client-side/javascript#initialize-the-client for instructions on SDK initialization.
console_appender.ts:31 [2024-08-19T19:48:17.140+02:00][ERROR][feature-flags-service.open-feature] Error running launchdarkly-client-provider's context change handler: Invalid context specified. Please see https://docs.launchdarkly.com/sdk/client-side/javascript#initialize-the-client for instructions on SDK initialization.
ldclient.cjs.js:1 debug: [LaunchDarkly] sending 1 events
</pre>
</details>

Reply:
Great, thanks for trying all these configurations! Everything looks safe to me, but I'm not sure about these lines:

> feature_flags_service.ts:155 GET https://app.launchdarkly.com/sdk/evalx/668ea35e5dd98d109e1bc7d/contexts/eyJraW5kIjoibXVsdGkiLCJraWJhbmEiOnsia2V5IjoiMTIzNCIsIm9mZmVyaW5nIjoidHJhZGl0aW9uYWwiLCJ2ZXJzaW9uIjoiOC4xNi4wIiwiYnVpbGRfbnVtIjo5MDA3MTk5MjU0NzQwOTkxLCJidWlsZF9zaGEiOiJYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYIiwiYnVpbGRfc2hhX3Nob3J0IjoiWFhYWFhYWFhYWFhYIn0sIm9yZ2FuaXphdGlvbiI6eyJrZXkiOiIxMjM0NSIsImlzX2VsYXN0aWNfc3RhZmYiOnRydWV9fQ 404 (Not Found)

> ldclient.cjs.js:1 debug: [LaunchDarkly] polling for feature flags at https://app.launchdarkly.com/sdk/evalx/668ea35e5dd98d109e1bc7d0/contexts/eyJhbm9ueW1vdXMiOnRydWUsImtleSI6ImJmYjBiMjYwLTQzOGUtMTFlZi05OGJmLTFmNTk2ODdjY2ZmOCJ9

> ldclient.cjs.js:1 info: [LaunchDarkly] Opening stream connection to https://clientstream.launchdarkly.com/eval/668ea35e5dd98d109e1bc7d0/eyJraW5kIjoibXVsdGkiLCJraWJhbmEiOnsia2V5IjoiMTIzNCIsIm9mZmVyaW5nIjoidHJhZGl0aW9uYWwiLCJ2ZXJzaW9uIjoiOC4xNi4wIiwiYnVpbGRfbnVtIjo5MDA3MTk5MjU0NzQwOTkxLCJidWlsZF9zaGEiOiJYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYIiwiYnVpbGRfc2hhX3Nob3J0IjoiWFhYWFhYWFhYWFhYIiwiaGFzX2RhdGEiOmZhbHNlfSwib3JnYW5pemF0aW9uIjp7ImtleSI6IjEyMzQ1IiwiaXNfZWxhc3RpY19zdGFmZiI6dHJ1ZX19

These are [JWTs in the URLs](https://jwt.io/#debugger-io?token=eyJraW5kIjoibXVsdGkiLCJraWJhbmEiOnsia2V5IjoiMTIzNCIsIm9mZmVyaW5nIjoidHJhZGl0aW9uYWwiLCJ2ZXJzaW9uIjoiOC4xNi4wIiwiYnVpbGRfbnVtIjo5MDA3MTk5MjU0NzQwOTkxLCJidWlsZF9zaGEiOiJYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYIiwiYnVpbGRfc2hhX3Nob3J0IjoiWFhYWFhYWFhYWFhYIiwiaGFzX2RhdGEiOmZhbHNlfSwib3JnYW5pemF0aW9uIjp7ImtleSI6IjEyMzQ1IiwiaXNfZWxhc3RpY19zdGFmZiI6dHJ1ZX0sImFsZyI6IkhTMjU2In0.e30.SNGtihUaAgHTiDj_A8WNFdBP5B4raiG6OZTiBwh0Z-Y). Do we treat them as non-secrets and are okay with exposing them in our logs?

![image](https://github.com/user-attachments/assets/3a4c9cd5-1524-462b-8cb6-d86ef0109661)

Reply:
IMO, the information shipped to LaunchDarkly should never include secrets. After all, it's a 3rd-party, and we don't have full control of how they treat the data. Do you think anything listed in the context should be considered a secret?
If we ever consider sending anything potentially problematic, we can use [private attributes](https://docs.launchdarkly.com/home/observability/private-context-attributes).

OTOH, those URLs are also visible in the Browser's Network tab, correct? Logging them in case of an error is not a big difference. If we want to remove the context from the URL, LaunchDarkly offers [the option `useReport: true`](https://docs.launchdarkly.com/guides/account/user-data/?q=user+data#how-to-minimize-end-user-data-sent-to-launchdarkly): it uses the `REPORT` verb and ships the context in the body.
TBH, I was surprised that, while it initially looks like a preferred option, it's not enabled by default. Then I noticed that the docs claim that [using this feature requires a custom polyfill for EventSource](https://docs.launchdarkly.com/sdk/client-side/javascript/requirements-polyfills#eventsource). The [docs of the polyfill](https://github.com/launchdarkly/js-eventsource) claim that it's not optimized for browsers 🤦 

Somehow, a custom polyfill to replace the native EventSource seems more risky to me.
WDYT?

Reply:
> Do you think anything listed in the context should be considered a secret?

What I see in those JWTs from the logs doesn't look sensitive to me, but I wasn't sure a) if that's everything they can contain or not, and b) what an adversary could do if they got access to this token. Is the JWT payload basically what's described as "context" in the RFC?

> If we ever consider sending anything potentially problematic, we can use [private attributes](https://docs.launchdarkly.com/home/observability/private-context-attributes).

Agreed.

> OTOH, those URLs are also visible in the Browser's Network tab, correct?

You tell me, I don't know 🙂 

> Logging them in case of an error is not a big difference.

If it's the case, then yeah, agree.

> Somehow, a custom polyfill to replace the native EventSource seems more risky to me.

++

---

All in all, I think my confusion/concern comes from not knowing much about these JWTs and not being sure what they're used for. But if they’re just carriers for the "context", then it should be fine.